src/memory-analyzer/gdb_api.cpp

```diff
@@ -0,0 +1,491 @@
+/*******************************************************************\
+
+Module: GDB Machine Interface API
+
+Author: Malte Mues <mail.mues@gmail.com>
+        Daniel Poetzl
+
+\*******************************************************************/
+
+/// \file
+/// Low-level interface to gdb
+
+// clang-format off
+#if defined(__linux__) || \
+    defined(__FreeBSD_kernel__) || \
+    defined(__GNU__) || \
+    defined(__unix__) || \
+    defined(__CYGWIN__) || \
+    defined(__MACH__)
+// clang-format on
+
+#include <cctype>
+#include <cerrno>
+#include <cstdio>
+#include <cstring>
+#include <regex>
+
+#include "gdb_api.h"
+
+#include <goto-programs/goto_model.h>
+
+#include <util/string_utils.h>
+
+/// Create a gdb_apit object
+///
+/// \param binary the binary to run with gdb
+/// \param log boolean indicating whether gdb input and output should be logged
+gdb_apit::gdb_apit(const char *binary, const bool log)
+  : binary(binary), log(log), gdb_state(gdb_statet::NOT_CREATED)
+{
+}
+
+/// Create a new gdb process for analysing the binary indicated by the member
+/// variable `binary`
+void gdb_apit::create_gdb_process()
+{
+  PRECONDITION(gdb_state == gdb_statet::NOT_CREATED);
+
+  command_log.clear();
+
+  pid_t gdb_process;
+
+  int pipe_input[2];
+  int pipe_output[2];
+
+  if(pipe(pipe_input) == -1)
+  {
+    throw gdb_interaction_exceptiont("could not create pipe for stdin");
+  }
+
+  if(pipe(pipe_output) == -1)
+  {
+    throw gdb_interaction_exceptiont("could not create pipe for stdout");
+  }
+
+  gdb_process = fork();
+
+  if(gdb_process == -1)
+  {
+    throw gdb_interaction_exceptiont("could not create gdb process");
+  }
+
+  if(gdb_process == 0)
+  {
+    // child process
+    close(pipe_input[1]);
+    close(pipe_output[0]);
+
+    dup2(pipe_input[0], STDIN_FILENO);
+    dup2(pipe_output[1], STDOUT_FILENO);
+    dup2(pipe_output[1], STDERR_FILENO);
+
+    dprintf(pipe_output[1], "binary name: %s\n", binary);
+
+    char *arg[] = {const_cast<char *>("gdb"),
+                   const_cast<char *>("--interpreter=mi"),
+                   const_cast<char *>(binary),
+                   NULL};
+
+    dprintf(pipe_output[1], "Loading gdb...\n");
+    execvp("gdb", arg);
+
+    // Only reachable, if execvp failed
+    int errno_value = errno;
+    dprintf(pipe_output[1], "errno in child: %s\n", strerror(errno_value));
+  }
+  else
+  {
+    // parent process
```

Comment:
Depending on how paranoid/careful you want to be, do you want to install a signal handler to catch SIGCHILD and maybe SIGPIPE to handle either GDB crashing, or a user/OS process killing gdb underneath you?

Reply:
We currently detect such errors when either a write or read to/from the gdb process fails, in which case we throw an exception. I'm not sure if there are any specific problems with this approach but for now it's probably sufficient.